swri_roscpp/include/swri_roscpp/parameters.h

```diff
@@ -126,8 +153,65 @@ namespace swri
       bool &variable,
       const bool default_value)
   {
+    std::string resolved_name = nh.resolveName(name);
+    _used_params.insert(resolved_name);
     nh.param(name, variable, default_value);
     ROS_INFO("Read parameter %s = %s", name.c_str(), variable ? "true" : "false");
   }
+  
+  inline int comparePrefix(std::string const& string, std::string const& prefix)
+  {
+	return string.compare(0, prefix.size(), prefix);
+  }
+
+  inline bool isPrefixOf(std::string const& prefix, std::string const& string)
+  {
+	return comparePrefix(string, prefix) == 0;
+  }
+
+  inline int prefixLessThan(std::string const& string, std::string const& prefix)
+  {
+	return string.compare(0, prefix.size(), prefix) < 0;
+  }
+  
+  /**
+   * Gets the list of parameters on the parameter server in the namespace of
+   * NodeHandle n that have not yet been got using any of the swri::param() or
+   * swri::getParam() functions.
+   *
+   * The parameter keys are filtered by prefix, so if there exists a parameter
+   * /foo/bar/baz and parameter /foo/bar has been got, then /foo/bar/baz will
+   * not be returned as an unused parameter.
+   *
+   * Note that this function has no way of knowing about parameters got
+   * using the ros::param() and ros::getParam() functions.
+   *
+   * @param n  Nodehandle defining the namespace to restrict the list
+   * @return   A vector of fully-qualified parameter names in n's
+   *           namespace that haven't been got.
+   */
+  std::vector<std::string> getUnusedParamKeys(ros::NodeHandle const& n)
+  {
+	std::string const& n_namespace = n.getNamespace();
+	std::vector<std::string> all_params;
+	n.getParamNames(all_params);
+	std::sort(all_params.begin(), all_params.end());
+	std::vector<std::string> unused_params(all_params.size());
+	boost::function<bool(std::string)> inNamespace = boost::bind(isPrefixOf, n_namespace, _1);
+	std::vector<std::string>::const_iterator first_param_in_namespace = find_if(all_params.begin(), all_params.end(), inNamespace);
+	std::vector<std::string>::const_iterator after_last_param_in_namespace = find_if(all_params.begin(), all_params.end(), std::not1(inNamespace));
+	std::vector<std::string>::iterator last_unused_param = std::set_difference(first_param_in_namespace, after_last_param_in_namespace, _used_params.begin(), _used_params.end(), unused_params.begin(), prefixLessThan);
+	unused_params.resize(last_unused_param - unused_params.begin());
+	return unused_params;
+  }
```

Comment:
How about we go simpler here with something like:

std::vector<std::string> result;
for (blahblah it = all_params.begin(); it != all_params.end(); ++it) {
  if (*it.compare(0, n_namespace.size(), n) == 0 && used_params.count(*it) == 0) {
    result.push_back(*it);
  }
}